File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/CommandsRadarTest.java
package dev.robocode.tankroyale.botapi;

import dev.robocode.tankroyale.schema.BotIntent;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import test_utils.MockedServer;

import java.util.function.Predicate;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests for radar commands (TR-API-CMD-003).
 */
@Tag("CMD")
@Tag("TR-API-CMD-003")
@DisplayName("Radar Commands (TR-API-CMD-003)")
class CommandsRadarTest extends AbstractBotTest {

    private static class RadarTestBot extends Bot {
        RadarTestBot() {
            super(botInfo, MockedServer.getServerUrl());
        }
    }

    private RadarTestBot startRadarBot() {
        var bot = new RadarTestBot();
        startAsync(bot);
        // We must wait for the game to start and the first tick (Turn 1).
        awaitGameStarted(bot);
        awaitTick(bot);

        // No need to drain Turn 1 here, we'll do it in awaitExpectedIntent if needed.
        return bot;
    }

    /**
     * Helper to wait for an intent that satisfies a predicate.
     * This handles draining multiple intents if the bot is looping automatically.
     */
    private void awaitExpectedIntent(Predicate<BotIntent> predicate) {
        long start = System.currentTimeMillis();
        int count = 0;
        while (System.currentTimeMillis() - start < 10000) {
            count++;
            server.continueBotIntent();
            if (server.awaitBotIntent(2000)) {
                BotIntent intent = server.getBotIntent();
                System.out.println("Received intent " + count + ": rescan=" + intent.getRescan() 
                    + ", adjRadarBody=" + intent.getAdjustRadarForBodyTurn()
                    + ", adjRadarGun=" + intent.getAdjustRadarForGunTurn());
                if (predicate.test(intent)) {
                    return;
                }
                server.resetBotIntentLatch();
            } else {
                System.out.println("Timed out waiting for intent " + count);
            }
        }
        assertThat(false).as("Timed out waiting for expected intent after " + count + " attempts").isTrue();
    }

    @Test
    @DisplayName("setRescan() sets the rescan flag in the intent")
    void testRescanIntent() {
        var bot = startRadarBot();

        bot.setRescan();
        awaitExpectedIntent(intent -> Boolean.TRUE.equals(intent.getRescan()));
    }

    @Test
    @DisplayName("rescan() blocking call sets the rescan flag in the intent")
    void testBlockingRescan() {
        var bot = startRadarBot();

        // Run rescan in a separate thread because it's blocking
        goAsync(bot::rescan);
        awaitExpectedIntent(intent -> Boolean.TRUE.equals(intent.getRescan()));
    }

    @Test
    @DisplayName("setAdjustRadarForBodyTurn(true) sets the flag in the intent")
    void testAdjustRadarBodyTrue() {
        var bot = startRadarBot();

        bot.setAdjustRadarForBodyTurn(true);
        awaitExpectedIntent(intent -> Boolean.TRUE.equals(intent.getAdjustRadarForBodyTurn()));
    }

    @Test
    @DisplayName("setAdjustRadarForBodyTurn(false) sets the flag in the intent")
    void testAdjustRadarBodyFalse() {
        var bot = startRadarBot();

        // Set to true first, then to false
        bot.setAdjustRadarForBodyTurn(true);
        awaitExpectedIntent(intent -> Boolean.TRUE.equals(intent.getAdjustRadarForBodyTurn()));

        bot.setAdjustRadarForBodyTurn(false);
        awaitExpectedIntent(intent -> Boolean.FALSE.equals(intent.getAdjustRadarForBodyTurn()));
    }

    @Test
    @DisplayName("setAdjustRadarForGunTurn(true) sets the flag in the intent")
    void testAdjustRadarGunTrue() {
        var bot = startRadarBot();

        bot.setAdjustRadarForGunTurn(true);
        awaitExpectedIntent(intent -> Boolean.TRUE.equals(intent.getAdjustRadarForGunTurn()));
    }

    @Test
    @DisplayName("setAdjustRadarForGunTurn(false) sets the flag in the intent")
    void testAdjustRadarGunFalse() {
        var bot = startRadarBot();

        // Set to true first, then to false
        bot.setAdjustRadarForGunTurn(true);
        awaitExpectedIntent(intent -> Boolean.TRUE.equals(intent.getAdjustRadarForGunTurn()));

        bot.setAdjustRadarForGunTurn(false);
        awaitExpectedIntent(intent -> Boolean.FALSE.equals(intent.getAdjustRadarForGunTurn()));
    }
}

File: bot-api/dotnet/test/src/CommandsRadarTest.cs
using System;
using System.Threading;
using NUnit.Framework;
using Robocode.TankRoyale.Schema;

namespace Robocode.TankRoyale.BotApi.Tests;

/// <summary>
/// Tests for radar commands (TR-API-CMD-003).
/// </summary>
[TestFixture]
[Category("CMD")]
[Property("ID", "TR-API-CMD-003")]
[Description("Radar Commands (TR-API-CMD-003)")]
public class CommandsRadarTest : AbstractBotTest
{
    private class RadarTestBot : Bot
    {
        private int _blockingRescanRequested;

        public RadarTestBot(Uri serverUrl) : base(BotInfo, serverUrl) { }

        public override void Run()
        {
            while (IsRunning)
            {
                if (Interlocked.Exchange(ref _blockingRescanRequested, 0) == 1)
                {
                    Rescan();
                }
                else
                {
                    Go();
                }
            }
        }

        public void RequestBlockingRescan() => Interlocked.Exchange(ref _blockingRescanRequested, 1);
    }

    private RadarTestBot StartRadarBot()
    {
        var bot = new RadarTestBot(Server.ServerUrl);
        StartAsync(bot);
        AwaitGameStarted(bot);
        AwaitTick(bot);

        // Drain the initial automatic intent so subsequent captures start cleanly.
        Server.ResetBotIntentEvent();
        Server.ContinueBotIntent();
        AwaitBotIntent();
        Server.ResetBotIntentEvent();

        return bot;
    }

    /// <summary>
    /// Helper to wait for an intent that satisfies a predicate.
    /// This handles draining multiple intents if the bot is looping automatically.
    /// </summary>
    private void AwaitExpectedIntent(Predicate<BotIntent> predicate)
    {
        var start = DateTime.Now;
        while ((DateTime.Now - start).TotalSeconds < 30)
        {
            Server.ContinueBotIntent();
            if (Server.AwaitBotIntent(CiWaitMs))
            {
                if (predicate(Server.BotIntent))
                {
                    return;
                }
                Server.ResetBotIntentEvent();
            }
        }
        Assert.Fail("Timed out waiting for expected intent");
    }

    [Test]
    [Description("SetRescan() sets the rescan flag in the intent")]
    public void TestRescanIntent()
    {
        var bot = StartRadarBot();

        bot.SetRescan();
        AwaitExpectedIntent(intent => intent.Rescan == true);
    }

    [Test]
    [Description("Rescan() blocking call sets the rescan flag in the intent")]
    public void TestBlockingRescan()
    {
        var bot = StartRadarBot();

        // Rescan must run on the bot's managed thread because it blocks until the next turn.
        bot.RequestBlockingRescan();
        AwaitExpectedIntent(intent => intent.Rescan == true);
    }

    [Test]
    [Description("AdjustRadarForBodyTurn sets the flag in the intent")]
    public void TestAdjustRadarBody()
    {
        var bot = StartRadarBot();

        bot.AdjustRadarForBodyTurn = true;
        AwaitExpectedIntent(intent => intent.AdjustRadarForBodyTurn == true);

        bot.AdjustRadarForBodyTurn = false;
        AwaitExpectedIntent(intent => intent.AdjustRadarForBodyTurn == false);
    }

    [Test]
    [Description("AdjustRadarForGunTurn sets the flag in the intent")]
    public void TestAdjustRadarGun()
    {
        var bot = StartRadarBot();

        bot.AdjustRadarForGunTurn = true;
        AwaitExpectedIntent(intent => intent.AdjustRadarForGunTurn == true);

        bot.AdjustRadarForGunTurn = false;
        AwaitExpectedIntent(intent => intent.AdjustRadarForGunTurn == false);
    }
}
